pulsar-common/src/main/java/org/apache/pulsar/common/util/CompletableFutureCancellationHandler.java

```diff
@@ -41,10 +41,16 @@
  * any "downstream" dependent futures. A cancellation or timeout that happens in any "upstream"
  * future will get handled.
  */
-class CompletableFutureCancellationHandler {
-    private volatile boolean cancelled;
+public class CompletableFutureCancellationHandler {
+    private enum CompletionStatus {
+        PENDING,
+        CANCELLED,
+        DONE
+    }
+    private volatile CompletionStatus completionStatus = CompletionStatus.PENDING;
     private volatile Runnable cancelAction;
     private final AtomicBoolean cancelHandled = new AtomicBoolean();
+    private boolean attached;
```

Comment:
it should be fine since `attached` is used only in `attachToFuture` method and that method is synchronized. The reason why I added the "attached" field is to prevent misuse of `CompletableFutureCancellationHandler` class. It is designed to be "attached" to a single CompletableFuture and it cannot be reused.

Reply:
@lhotari - you're right, it is correct as written. I had a small misunderstanding about the guarantees of the `synchronized` keyword. (When the method's lock is released, the updates to `attached` will be flushed to main memory and then when the method's lock is re-acquired, another thread will have a consistent view of the variable.) Sorry for the confusion!